File: leetgo/go/LCR 181.fan-zhuan-dan-ci-shun-xu-lcof/solution.go
// Created by k at 2024/05/16 16:36
// leetgo: dev
// https://leetcode.cn/problems/fan-zhuan-dan-ci-shun-xu-lcof/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func reverseMessage(message string) string {
	var ans string
	i := len(message) - 1
	for  i >= 0 && message[i] == ' ' {
		i--
	}
	j := i
	for i >= 0 {
		for i >= 0 && (message[i]) != ' ' {
			i--
		}
		ans += message[i+1 : j+1]
		ans += " "
		for i >= 0 && message[i] == ' ' {
			i--
		}
		j = i
	}
	if len(ans) > 0 && ans[len(ans)-1] == ' '{
		ans = ans[0:len(ans)-1]
	}
	return ans

}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	message := Deserialize[string](ReadLine(stdin))
	ans := reverseMessage(message)

	fmt.Println("\noutput:", Serialize(ans))
}

File: leetgo/cpp/LCR 181.fan-zhuan-dan-ci-shun-xu-lcof/solution.cpp
// Created by k at 2024/05/17 15:56
// leetgo: dev
// https://leetcode.cn/problems/fan-zhuan-dan-ci-shun-xu-lcof/

#include <bits/stdc++.h>
#include "LC_IO.h"
using namespace std;

// @lc code=begin

class Solution {
  public:
    string reverseMessage(string message)
    {

        int i, j;
        i = message.size() - 1;
        while (i >= 0 && message[i] == ' ') {
            i--;
        }
        j = i;
        while (i >= 0) {
            while (i >= 0 && message[i] != ' ') {
                i--;
            }
            ans.append(message.substr(i + 1, j - i));
            ans.append(" ");
            while (i >= 0 && message[i] == ' ') {
                i--;
            }
            j = i;
        }
        if (ans.size() > 0 && ans.back() == ' ') {
            ans = ans.substr(0, ans.size() - 1);
        }
        return ans;
    }
    string ans;
};

// @lc code=end

int main()
{
    ios_base::sync_with_stdio(false);
    stringstream out_stream;

    string message;
    LeetCodeIO::scan(cin, message);

    Solution *obj = new Solution();
    auto res      = obj->reverseMessage(message);
    LeetCodeIO::print(out_stream, res);
    cout << "\noutput: " << out_stream.rdbuf() << endl;

    delete obj;
    return 0;
}
